File: api/trs80_coco.cxx
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.mikekohn.net/
 * License: GPLv3
 *
 * Copyright 2014-2018 by Michael Kohn
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "api/trs80_coco.h"

#define CHECK_FUNC(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->trs80_coco_##funct##sig(); \
  }

#define CHECK_FUNC_CONST(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->trs80_coco_##funct##sig(const_val); \
  }

#define CHECK_FUNC_CONST_2(funct,sig) \
  if (strcmp(#funct#sig, function) == 0) \
  { \
    return generator->trs80_coco_##funct##sig(const_val1, const_val2); \
  }

int trs80_coco(JavaClass *java_class, Generator *generator, const char *method_name)
{
  CHECK_FUNC(clearScreenLores,)
  CHECK_FUNC(clearScreenMidres,)
  //CHECK_FUNC(setBackgroundColor,_I)
  CHECK_FUNC(setText,_II)
  CHECK_FUNC(setGraphicsMode,)
  CHECK_FUNC(setTextMode,)
  CHECK_FUNC(print,_X)

  CHECK_FUNC(plotLores,_III)
  CHECK_FUNC(plotMidres,_III)

  CHECK_FUNC(initVideoFlags,)
  CHECK_FUNC(waitForHsync,)
  CHECK_FUNC(waitForVsync,)

  CHECK_FUNC(initSound,)
  CHECK_FUNC(setSound,_aB)

  return -1;
}

int trs80_coco(JavaClass *java_class, Generator *generator, const char *method_name, int const_val)
{
  //CHECK_FUNC_CONST(setBackgroundColor,_I)

  return -1;
}

int trs80_coco(JavaClass *java_class, Generator *generator, const char *function, int const_val1, int const_val2)
{
  //CHECK_FUNC_CONST_2(someFunction,_II)

  return -1;
}

File: java/net/mikekohn/java_grinder/TRS80Coco.java
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.naken.cc/
 * License: GPL
 *
 * Copyright 2014-2016 by Michael Kohn
 *
 */

package net.mikekohn.java_grinder;

abstract public class TRS80Coco
{
  public static final int MODE_3 = 3;

  public static final int COLOR_BLACK = 0x80;
  public static final int COLOR_GREEN = 0x8f;
  public static final int COLOR_YELLOW = 0x9f;
  public static final int COLOR_BLUE = 0xaf;
  public static final int COLOR_RED = 0xbf;
  public static final int COLOR_WHITE = 0xcf;
  public static final int COLOR_CYAN = 0xdf;
  public static final int COLOR_PURPLE = 0xef;
  public static final int COLOR_ORANGE = 0xff;

  protected TRS80Coco() { }

  public static void clearScreenLores() { }
  public static void clearScreenMidres() { }
  //public static void setBackgroundColor(int color) { }
  public static void setText(int address, int value) { }
  public static void setGraphicsMode() { }
  public static void setTextMode() { }
  public static void print(String text) { }

  /** Plot a colored block.  x can be between 0 and 31, y between 0 and 15 */
  public static void plotLores(int x, int y, int color) { }

  /** Plot a mid-res block (text mode).  x can be between 0 and 63 and y
      can be between 0 and 31 */
  public static void plotMidres(int x, int y, int color) { }

  public static void initVideoFlags() { }
  //public static void disableVsyncListener() { }
  public static void waitForHsync() { }
  public static void waitForVsync() { }

  /** Must be called before sound is used. */
  public static void initSound() { }

  /** Play the passed-in waveform through the 6 bit DAC or if NULL
      turn off sound */
  public static void setSound(byte[] waveform) { }
}

